### DISCUSSION TITLE
UnboundLocalError: cannot access local variable 'output'

### DISCUSSION THREAD

**Discussion by rvanmaaren** [semiautomaticgit/SemiAutomaticClassificationPlugin#358] [General]
Hi,

An error occurred while running the classification tool.

An error has occurred while executing Python code: 

UnboundLocalError: cannot access local variable 'output' where it is not associated with a value 
Traceback (most recent call last):
  File "C:\Users/Robin/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\SemiAutomaticClassificationPlugin\interface\classification_tab.py", line 143, in run_classification_action
    run_classifier()
  File "C:\Users/Robin/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\SemiAutomaticClassificationPlugin\interface\classification_tab.py", line 487, in run_classifier
    if output is None:
       ^^^^^^
UnboundLocalError: cannot access local variable 'output' where it is not associated with a value

Python version: 3.12.3 (main, Apr 14 2024, 17:21:43) [MSC v.1938 64 bit (AMD64)] 
QGIS version: 3.36.3-Maidenhead Maidenhead, 2df96554 

Python Path:
C:/PROGRA~1/QGIS33~1.3/apps/qgis/./python
C:/Users/Robin/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/Robin/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins
C:/PROGRA~1/QGIS33~1.3/apps/qgis/./python/plugins
C:\PROGRA~1\QGIS33~1.3\apps\grass\grass83\etc\python
C:\Users\Robin\Documents
C:\Program Files\QGIS 3.36.3\bin\python312.zip
C:\PROGRA~1\QGIS33~1.3\apps\Python312\DLLs
C:\PROGRA~1\QGIS33~1.3\apps\Python312\Lib
C:\Program Files\QGIS 3.36.3\bin
C:\PROGRA~1\QGIS33~1.3\apps\Python312
C:\PROGRA~1\QGIS33~1.3\apps\Python312\Lib\site-packages
C:\PROGRA~1\QGIS33~1.3\apps\Python312\Lib\site-packages\win32
C:\PROGRA~1\QGIS33~1.3\apps\Python312\Lib\site-packages\win32\lib
C:\PROGRA~1\QGIS33~1.3\apps\Python312\Lib\site-packages\Pythonwin
C:/Users/Robin/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:\Users\Robin\AppData\Roaming\QGIS\QGIS3\profiles\default\python\plugins
X:/Leden/5180 Provincie Noord Brabant/23335180 meerwaarden autochtone bomen en struiken

**Comment by semiautomaticgit** [comment]
Hi @rvanmaaren , 

it is the same issue https://github.com/semiautomaticgit/SemiAutomaticClassificationPlugin/issues/330

Please enable "Records detailed events in a log file" and copy the log file as described here https://semiautomaticclassificationmanual.readthedocs.io/en/latest/faq.html#how-can-i-report-an-error

**Comment by rvanmaaren** [comment]
Hi,

log file:
INFO|2024-07-08T09:22:25.629|system_tools|get_system_info|38|version:0.3.5; system: Windows; 64bit: True; n_processes: 2; ram: 10; temp.dir: C:\Users\Robin\AppData\Local\Temp\remotior_sensus_0708g2bv1arj
WARNING|2024-07-08T09:22:25.631|session|_check_dependencies|468|No module named 'torch'
WARNING|2024-07-08T09:22:25.632|session|_check_dependencies|473|No module named 'sklearn'

kind regards,

**Comment by rvanmaaren** [comment]
update:

I've downloaded an older version of SCP (7.10.11, randomly selected) and now it works. Something in the code may have changed that is causing the error. I also have gotten a covariance matrix error a couple of times. But there is no log on that. 

Kind regards

**Reply by semiautomaticgit** [reply]
Thank you for your feedback. I'll look into it

**Reply by rvanmaaren** [reply]
update:
I've updated QGIS to 3.8.0. Here version 7.10.11 does not work anymore. When creating a new training input the next error message occurs: "Information [4]: Select a raster; raster is not. loaded".

**Reply by semiautomaticgit** [reply]
Hello @rvanmaaren , I'm going to update the plugin on the 3rd of August and it should solve your issue.

**Comment by semiautomaticgit** [comment]
Hello @rvanmaaren , it was fixed in SCP version 8.3

**Reply by rvanmaaren** [reply]
Hi, I wanted to test the fixed plugin but I get the following error message: 
- Semi-Automatic Classification Plugin: Warning. Python library Remotior Sensus is outdated.This could cause errors, please update Remotior Sensus.

I'm using QGIS 3.38.1. and scp 8.3.0

**Reply by semiautomaticgit** [reply]
Hi, please uninstall the plugin and follow these installation steps https://semiautomaticclassificationmanual.readthedocs.io/en/latest/installation.html

**Reply by rvanmaaren** [reply]
Hi, thanks for the quick response. 

It would be helpful if users can manage the Python library versions within the plugin. A shell installation is for most people a scary thing to do. Could it not install/update the libraries at the installation of the plug-in?

Besides that, another error popped up. I've gotten this error before:

2024-08-19T14:01:00     CRITICAL    SCP : covariance matrix
2024-08-19T14:01:01     CRITICAL    SCP : multiprocess: error classification
2024-08-19T14:01:01     CRITICAL    Error : Process failed
 
The error occours when I'm running the classification tool. I've loaded in the bands and set up a training dataset.

**Reply by semiautomaticgit** [reply]
You are welcome.

> It would be helpful if users can manage the Python library versions within the plugin. A shell installation is for most people a scary thing to do. Could it not install/update the libraries at the installation of the plug-in?

Unfortunately, it is not possible to automatically install the libraries for security reasons.

> Besides that, another error popped up. I've gotten this error before:
> 
> 2024-08-19T14:01:00 CRITICAL SCP : covariance matrix 2024-08-19T14:01:01 CRITICAL SCP : multiprocess: error classification 2024-08-19T14:01:01 CRITICAL Error : Process failed
> 
> The error occours when I'm running the classification tool. I've loaded in the bands and set up a training dataset.

You can try a different algorithm.
Please copy the log file as described here https://semiautomaticclassificationmanual.readthedocs.io/en/latest/faq.html#how-can-i-report-an-error

**Reply by rvanmaaren** [reply]
I have done the same classification as before. It has worked for the test data I'm using in an older version of QGIS and an older version of SCP. But when I'm using the newest version of QGIS and scp it does not work. I use the band processing classification tool and the default to run.

2024-08-21T07:42:00     CRITICAL    SCP : covariance matrix
2024-08-21T07:42:01     CRITICAL    SCP : multiprocess: error classification
2024-08-21T07:42:01     CRITICAL    Error : Process failed

[debugfile.txt](https://github.com/user-attachments/files/16686463/debugfile.txt)

**Reply by semiautomaticgit** [reply]
Thank you for the log file.
It seems an issue with input data.
You can try to rename the input raster to a shorter name starting with a letter (e.g. Sentinel2.tif).
Also, you can try to split the raster to single bands using this tool https://semiautomaticclassificationmanual.readthedocs.io/en/latest/split_bands_tab.html and use these bands in a band set to perform the classification.

**Reply by rvanmaaren** [reply]
Strangely in previous versions the classification tool did work with the input data. After renaming the data I got the same error. After splitting the bands the classification tool worked. Its a bit of a workaround but it works. Thanks for the help :)

[debugfile2.txt](https://github.com/user-attachments/files/16691145/debugfile2.txt)

**Reply by semiautomaticgit** [reply]
You are welcome, I'm glad it worked.
I'll try to understand if there are issues with multiband rasters.
